### PR TITLE
Define output files' format. Add `OutputManager` & `PostProcessManager`

### PR DESCRIPTION
**Pull request by anlavandier**
Define format of output files:
- A single YAML file contains the information of all the FEM spaces;
- A single HDF5 file contains the coefficients of all the FEM fields (both static and time dependent).

New classes:
- `OutputManager` exports FEM spaces and fields from current simulation to output files;
- `PostProcessManager` loads data from geometry and output files, evaluates fields on user-defined grid, and exports to VTK.

In order to avoid excessive slowdown of operations in `PostProcessManager`, fast evaluation methods (accelerated with Pyccel) were added to the classes `TensorFemSpace`, `ProductFemSpace` and `VectorFemSpace`.

### CODE REVIEW

**Review by yguclu — COMMENTED** [review]
I do not think that the `output-dev` directory should be added to `devel`. Do you want to remove it now, or prefer to keep it for reference until the PR is ready for merging?

**Comment by anlavandier** [comment]
> I do not think that the `output-dev` directory should be added to `devel`. Do you want to remove it now, or prefer to keep it for reference until the PR is ready for merging?

I think it's better to wait until it's ready to merge to remove it.

**Comment by anlavandier** [comment]
i think we should wait until #193 is merged so I can fix the numerous conflicts that will arise from this.

**Comment by yguclu** [comment]
PR #193 has been merged, please resolve the conflicts with `devel` now.

File: psydac/api/postprocessing.py
Comment (yguclu) [inline psydac/api/postprocessing.py:380]: In classical Pythonic fashion, thanks to short-circuiting of the boolean operators, you can just write
```python
if comm and comm.size > 1:
```
because in the case where `comm is None` the Python interpreter knows that the logical expression is `False` and stops before evaluating `comm.size > 1`.

File: psydac/api/postprocessing.py
Comment (yguclu) [inline psydac/api/postprocessing.py:355]: What do you mean?

**Review by yguclu — CHANGES_REQUESTED** [review]
Excellent work @anlavandier!

I have two comments:

1. Many functions in `psydac.core.kernels` do not appear to be tested. I think they should be, directly with some unit tests (preferable), or indirectly through a more complex API test.
2. We need to provide a brief documentation of our data file format, and of the new classes. Maybe we can start with a short Markdown file in a new `documentation` directory?

File: psydac/api/tests/test_post_processing.py
Comment (anlavandier) [reply psydac/api/tests/test_post_processing.py:221]: Well, you could always look at the `vtk` files to see that they look right (we would have to change the file slightly to write recognizable data and not just random values) or in the case of the identity mappings, check that the `pushforward` function gives us the same values. In the `bent_pipe` case I don't see how it could be checked without looking that the `vtk` file.

File: psydac/core/kernels.py
Comment (anlavandier) [reply psydac/core/kernels.py:453]: The generated code I looked at also did and when I copied chunks from it I ordered them in the same way.

File: psydac/api/postprocessing.py
Comment (anlavandier) [reply psydac/api/postprocessing.py:96]: It does. https://www.merriam-webster.com/dictionary/exportation

File: psydac/api/postprocessing.py
Comment (anlavandier) [reply psydac/api/postprocessing.py]: Fixed in 0c5bd9b

File: psydac/api/postprocessing.py
Comment (anlavandier) [reply psydac/api/postprocessing.py:139]: Done in 0c5bd9b

File: psydac/api/postprocessing.py
Comment (anlavandier) [reply psydac/api/postprocessing.py:67]: It looked weird so I added a space to have be the same as the two previous lines.

File: psydac/core/kernels.py
Comment (anlavandier) [reply psydac/core/kernels.py]: Done in 0c5bd9b

File: psydac/core/kernels.py
Comment (anlavandier) [reply psydac/core/kernels.py]: Done in 0c5bd9b

File: psydac/core/kernels.py
Comment (anlavandier) [reply psydac/core/kernels.py]: Done in 0c5bd9b

File: psydac/core/kernels.py
Comment (anlavandier) [reply psydac/core/kernels.py]: Done in 0c5bd9b

File: psydac/core/kernels.py
Comment (anlavandier) [reply psydac/core/kernels.py]: Done in 0c5bd9b

File: psydac/core/kernels.py
Comment (anlavandier) [reply psydac/core/kernels.py]: Done in 0c5bd9b

File: psydac/mapping/discrete.py
Comment (anlavandier) [reply psydac/mapping/discrete.py]: Done in 0c5bd9b

Comment (anlavandier) [reply psydac/mapping/discrete.py]: Done in 843f7ad

File: psydac/fem/tensor.py
Comment (anlavandier) [reply psydac/fem/tensor.py]: Done in 0c5bd9b

File: psydac/fem/vector.py
Comment (anlavandier) [reply psydac/fem/vector.py]: Done in 0c5bd9b

File: psydac/fem/tensor.py
Comment (anlavandier) [reply psydac/fem/tensor.py]: Done in 0c5bd9b

File: psydac/fem/vector.py
Comment (anlavandier) [reply psydac/fem/vector.py]: Done in 0c5bd9b

File: psydac/fem/vector.py
Comment (anlavandier) [reply psydac/fem/vector.py]: Done in 0c5bd9b

File: psydac/fem/vector.py
Comment (anlavandier) [reply psydac/fem/vector.py]: Done in 0c5bd9b

File: psydac/fem/vector.py
Comment (anlavandier) [reply psydac/fem/vector.py]: Done in 0c5bd9b

File: psydac/fem/vector.py
Comment (anlavandier) [reply psydac/fem/vector.py]: Done in 0c5bd9b

File: psydac/fem/splines.py
Comment (anlavandier) [reply psydac/fem/splines.py]: Done in 0c5bd9b

File: psydac/core/tests/test_bsplines.py
Comment (anlavandier) [reply psydac/core/tests/test_bsplines.py]: Done in [dbb89a4](https://github.com/pyccel/psydac/pull/165/commits/dbb89a4608ce1b3798465149612ff37e4504ad14)

File: psydac/api/tests/test_post_processing.py
Comment (anlavandier) [reply psydac/api/tests/test_post_processing.py]: Done in https://github.com/pyccel/psydac/commit/0c5bd9b1ca6859eeb1ceed670e579532b56833b4

File: psydac/api/tests/test_post_processing.py
Comment (anlavandier) [reply psydac/api/tests/test_post_processing.py]: Done in https://github.com/pyccel/psydac/commit/0c5bd9b1ca6859eeb1ceed670e579532b56833b4

File: psydac/mapping/discrete.py
Comment (anlavandier) [reply psydac/mapping/discrete.py]: Removed in https://github.com/pyccel/psydac/commit/0c5bd9b1ca6859eeb1ceed670e579532b56833b4

File: psydac/api/postprocessing.py
Comment (anlavandier) [reply psydac/api/postprocessing.py]: removed in [843f7ad](https://github.com/pyccel/psydac/pull/165/commits/843f7ad47f8a3a3c8938f42682631c9231959f9d)

File: psydac/fem/vector.py
Comment (yguclu) [reply psydac/fem/vector.py:600]: Yes, but this is not a problem given that `symbolic_space` is a SymPDE object:
```python
from sympde.topology.space import BasicFunctionSpace
```
I think that the other spaces are defined in the same module.

File: psydac/core/kernels.py
Comment (yguclu) [reply psydac/core/kernels.py:453]: I see 😅 . Then I probably know why, as these instructions are successively inserted at the first location of a block of multiple lines of code. It was probably easier to do it this way.

Please order them in a more human/natural way, if you don't mind...

File: psydac/api/postprocessing.py
Comment (yguclu) [reply psydac/api/postprocessing.py:96]: I stand corrected 😅

To my defense, it's the first time I hear it. I've always used the noun "export" instead. 😆

File: psydac/api/postprocessing.py
Comment (yguclu) [reply psydac/api/postprocessing.py]: I see.

I do not recall our discussion in every detail, but I imagine that there are situations where the user may want to export many fields, and maybe they are not all available at the same location in the code. So in this case it could be convenient to call `export_fields` multiple times on the same snapshot...

File: psydac/mapping/discrete.py
Comment (yguclu) [reply psydac/mapping/discrete.py]: All right

File: psydac/core/kernels.py
Comment (yguclu) [reply psydac/core/kernels.py:1316]: Perfect, there should be no temporary

File: psydac/api/postprocessing.py
Comment (yguclu) [reply psydac/api/postprocessing.py:67]: Fine

**Comment by anlavandier** [comment]
> Two comments:
> 1.  The "determinant of the Jacobian matrix" is simply called the "Jacobian determinant";
> 2.   In this case we are computing something else, namely the "determinant of the metric matrix" which is known as "metric determinant".

I was actually computing the jacobian. I have changed it compute the metric determinant and update the docstrings in 073d8b1 and 4294080

File: psydac/core/kernels.py
Comment (anlavandier) [reply psydac/core/kernels.py]: Fixed in 073d8b1

File: psydac/fem/vector.py
Comment (anlavandier) [reply psydac/fem/vector.py]: Fixed in d33d909

**Comment by yguclu** [comment]
> > Two comments:
> > 
> > 1. The "determinant of the Jacobian matrix" is simply called the "Jacobian determinant";
> > 2. In this case we are computing something else, namely the "determinant of the metric matrix" which is known as "metric determinant".
> 
> I was actually computing the Jacobian determinant. I have changed it compute the metric determinant and update the docstrings in [073d8b1](https://github.com/pyccel/psydac/commit/073d8b17c84e2e7adaec398f2d3cc27f622ddbb9) and [4294080](https://github.com/pyccel/psydac/commit/429408023d5fe2d387e6993a72e3f761dc8a2321) and [cc4a393](https://github.com/pyccel/psydac/commit/cc4a39348197061a8beaff9ceb98450354023324)

Depending on the use case, it might be OK to compute the Jacobian determinant instead of the metric one! Which one do you use in practice for the push-forward operations?

**Comment by anlavandier** [comment]
> 1. Many functions in `psydac.core.kernels` do not appear to be tested. I think they should be, directly with some unit tests (preferable), or indirectly through a more complex API test.
>
> 2. We need to provide a brief documentation of our data file format, and of the new classes. Maybe we can start with a short Markdown file in a new `documentation` directory?

For 2., maybe it is the right time to start to incorporate automatically generated documentation from the dosctrings. 

> > > Two comments:
> > > 
> > > 1. The "determinant of the Jacobian matrix" is simply called the "Jacobian determinant";
> > > 2. In this case we are computing something else, namely the "determinant of the metric matrix" which is known as "metric determinant".
> > 
> > 
> > I was actually computing the Jacobian determinant. I have changed it compute the metric determinant and update the docstrings in [073d8b1](https://github.com/pyccel/psydac/commit/073d8b17c84e2e7adaec398f2d3cc27f622ddbb9) and [4294080](https://github.com/pyccel/psydac/commit/429408023d5fe2d387e6993a72e3f761dc8a2321) and [cc4a393](https://github.com/pyccel/psydac/commit/cc4a39348197061a8beaff9ceb98450354023324)
> 
> Depending on the use case, it might be OK to compute the Jacobian determinant instead of the metric one! Which one do you use in practice for the push-forward operations?

I use the square root of the metric determinant to pushforward on a L2 mapping. The reason I use that is because it's what is done with the `CallableMapping` of SymPDE in `psydac/feec/pull_push.py`

**Comment by yguclu** [comment]
> > 1. Many functions in `psydac.core.kernels` do not appear to be tested. I think they should be, directly with some unit tests (preferable), or indirectly through a more complex API test.
> > 2. We need to provide a brief documentation of our data file format, and of the new classes. Maybe we can start with a short Markdown file in a new `documentation` directory?
> 
> For 2., maybe it is the right time to start to incorporate automatically generated documentation from the dosctrings.
> 
> > > > Two comments:
> > > > 
> > > > 1. The "determinant of the Jacobian matrix" is simply called the "Jacobian determinant";
> > > > 2. In this case we are computing something else, namely the "determinant of the metric matrix" which is known as "metric determinant".
> > > 
> > > 
> > > I was actually computing the Jacobian determinant. I have changed it compute the metric determinant and update the docstrings in [073d8b1](https://github.com/pyccel/psydac/commit/073d8b17c84e2e7adaec398f2d3cc27f622ddbb9) and [4294080](https://github.com/pyccel/psydac/commit/429408023d5fe2d387e6993a72e3f761dc8a2321) and [cc4a393](https://github.com/pyccel/psydac/commit/cc4a39348197061a8beaff9ceb98450354023324)
> > 
> > 
> > Depending on the use case, it might be OK to compute the Jacobian determinant instead of the metric one! Which one do you use in practice for the push-forward operations?
> 
> I use the square root of the metric determinant to pushforward on a L2 mapping. The reason I use that is because it's what is done with the `CallableMapping` of SymPDE in `psydac/feec/pull_push.py`. The already written point by point API of the `SplineMapping` class also gives the metric determinant.

Unfortunately the situation is a bit complicated, there is probably a sign missing (which is either `+1` or `-1`) in SymPDE: see [here](https://github.com/pyccel/psydac/blob/de19fd5856d8984927720c989de7c0bcce295238/psydac/feec/pull_push.py#L399-L410)

**Review by yguclu — CHANGES_REQUESTED** [review]
Excellent work @anlavandier!

Can you please add an example of usage for class `PostProcessManager`?

File: output.md
Comment (yguclu) [inline output.md:100]: Typo: and the -> all the

File: output.md
Comment (yguclu) [inline output.md:102]: I think it is much clearer to show two separate code snippets, rather than saying "The `#` line is supposed to represent the end of a file."

**Review by yguclu — CHANGES_REQUESTED** [review]

File: output.md
Comment (anlavandier) [reply output.md]: Fixed in 9d1e1b5

File: output.md
Comment (anlavandier) [reply output.md]: Fixed in 9d1e1b5

File: output.md
Comment (yguclu) [reply output.md]: Thanks. I have just expanded the explanations

**Review by yguclu — APPROVED** [review]

**Pull request merged**
